src/main/java/org/folio/circulation/domain/validation/ItemLimitValidator.java

```diff
@@ -0,0 +1,156 @@
+import java.util.function.Function;
+
+import io.vertx.core.json.JsonObject;
+import io.vertx.ext.web.RoutingContext;
+import org.apache.commons.lang3.StringUtils;
+
+import org.folio.circulation.domain.Item;
+import org.folio.circulation.domain.Loan;
+import org.folio.circulation.domain.LoanAndRelatedRecords;
+import org.folio.circulation.domain.LoanRepository;
+import org.folio.circulation.domain.policy.LoanPolicyRepository;
+import org.folio.circulation.support.CollectionResourceClient;
+import org.folio.circulation.support.Result;
+import org.folio.circulation.support.ValidationErrorFailure;
+
+public class ItemLimitValidator {
+  private final Function<String, ValidationErrorFailure> itemLimitErrorFunction;
+  private final LoanRepository loanRepository;
+  private final LoanPolicyRepository loanPolicyRepository;
+  private final CollectionResourceClient circulationRulesStorage;
+  private final RoutingContext routingContext;
+
+  public ItemLimitValidator(Function<String, ValidationErrorFailure> itemLimitErrorFunction,
+    LoanRepository loanRepository, LoanPolicyRepository loanPolicyRepository,
+    CollectionResourceClient circulationRulesStorage, RoutingContext routingContext) {
+
+    this.itemLimitErrorFunction = itemLimitErrorFunction;
+    this.loanRepository = loanRepository;
+    this.loanPolicyRepository = loanPolicyRepository;
+    this.circulationRulesStorage = circulationRulesStorage;
+    this.routingContext = routingContext;
+  }
+
+  public CompletableFuture<Result<LoanAndRelatedRecords>> refuseWhenItemLimitIsReached(
+    LoanAndRelatedRecords records) {
+
+    Loan loan = records.getLoan();
+    Integer itemLimit = loan.getLoanPolicy().getItemLimit();
+
+    if (itemLimit == null) {
+      return completedFuture(succeeded(records));
+    }
+
+    return loanPolicyRepository.lookupLineNumber(loan.getItem(), loan.getUser())
+      .thenComposeAsync(r -> r.after(this::retrieveCirculationRule)
+        .thenComposeAsync(result -> result.failAfter(rule -> isLimitReached(rule, records),
+          rule -> {
+            String message = getErrorMessage(rule);
+            return itemLimitErrorFunction.apply(String.format("Patron has reached maximum item limit of %d items %s",
+              itemLimit, message));
+          }))
+        .thenApply(result -> result.map(v -> records)));
+  }
+
+  private CompletableFuture<Result<Boolean>> isLimitReached(String rule, LoanAndRelatedRecords records) {
+
+    if (!isRuleMaterialTypePresent(rule) && !isRuleLoanTypePresent(rule)) {
+      return ofAsync(() -> false);
+    }
+
+    Item item = records.getLoan().getItem();
+    String materialType = item.getMaterialType().getString("name");
+    String loanType = item.getLoanTypeName();
+    Integer itemLimit = records.getLoan().getLoanPolicy().getItemLimit();
+
+    if (isRuleMaterialTypePresent(rule) || isRuleLoanTypePresent(rule)) {
+      return loanRepository.findOpenLoansByUserIdAndLoanPolicyIdWithItem(records)
```

Comment:
Why are the loans filtered by loan policy ID? A loan for an item with the same material type and loan type might have used a different loan policy?

Reply:
I followed the next logic: 

- Determine loan policy based on circulation rules (already exists);
- Get rule conditions for current circulation rule; 
- If rule contains loan type and/or material type - to filter retrieved list of loans by material type and loan type, if not - count all records. Then I compare the count of loans with item limit of the loan policy. I use the loan policy id  as a criteria to find loans (open status + userId + loanPolicyId).

Reply:
Why is the loan policy included in the criteria for finding loans? I don't understand how that fits into the requirements.

Reply:
We define item limit for specific loan policy. But this loan policy can be used with another material types or loan types. So if material types or loan types are present in circulation rule we should filter the result, if not - count records. But without loan policy ID we can't find records with appropriate policy.

Reply:
I don't follow, why is the loan policy ID needed to find the the relevant items? 

You stated above that the items are filtered by the material type and/or loan type, why is the loan policy relevant for filtering?

Reply:
But... if we didn't have policy ID, how would we check that fetched loans match the policy with specific item limit? Could you please clarify?

Reply:
As I understand it, the loans do not have to match the policy.

The limit is based upon a count of open loans for items with a matching material type and/or loan type in the rule that determined which policy to use.

Once the rules have be applied, both the criteria of the rule and the loan policy (and hence limit) are known. Finding the relevant items does not require the loan policy ID at all.

I believe filtering the loans by the loan policy ID could result in incorrect results.

Does that make sense?

Reply:
I have removed loan policy ID criteria from the query.
Thanks, updated

Reply:
@roman-barannyk Thanks